### PR TITLE
CEDS-2906 MRN validation

### PR DESCRIPTION
**Pull request by MaciejRewera**
This PR constitutes changes required to validate
MRN entered by the user by fetching Declaration
Status from Customs Declarations Information 
service. It is further checked if MRN and EORI 
from fetched document are the same as the ones
used by the user.

### CODE REVIEW

File: test/connectors/CdsFileUploadConnectorSpec.scala [hunk 1]
```diff
@@ -34,23 +36,27 @@ class CdsFileUploadConnectorSpec extends UnitSpec with BeforeAndAfterEach with I
...
   val hc = mock[HeaderCarrier]
 
-  override protected def afterEach(): Unit = {
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
     reset(httpClient)
+  }
 
+  override protected def afterEach(): Unit = {
+    reset(httpClient)
```
Comment (glidester) [inline test/connectors/CdsFileUploadConnectorSpec.scala:45]: Do we need to reset both before and after each test is run?

File: test/connectors/CdsFileUploadConnectorSpec.scala [hunk 2]
```diff
@@ -72,4 +78,46 @@ class CdsFileUploadConnectorSpec extends UnitSpec with BeforeAndAfterEach with I
...
+
+        val httpResponse = HttpResponse(status = NOT_FOUND, body = "")
+
+        when(httpClient.GET[HttpResponse](anyString())(any(), any(), any()))
+          .thenReturn(Future.successful(httpResponse))
+
+        val result = cdsFileUploadConnector.getDeclarationStatus(MRN(CommonTestData.mrn).get)(hc).futureValue
+
+        result mustBe httpResponse
+      }
```
Comment (glidester) [inline test/connectors/CdsFileUploadConnectorSpec.scala:120]: Should we also test that any 'other' status throws the expected exception?

File: test/controllers/MrnEntryControllerSpec.scala [hunk 3]
```diff
@@ -18,108 +18,136 @@ package controllers
...
-    when(page.apply(any())(any(), any())).thenReturn(HtmlFormat.empty)
+    reset(mrnEntryPage, mrnAccessDeniedPage, mrnDisValidator)
+    when(mrnEntryPage.apply(any())(any(), any())).thenReturn(HtmlFormat.empty)
+    when(mrnAccessDeniedPage.apply(any())(any(), any())).thenReturn(HtmlFormat.empty)
   }
 
-  override protected def afterEach(): Unit = {
-    reset(page)
+  override def afterEach(): Unit = {
+    reset(mrnEntryPage, mrnAccessDeniedPage, mrnDisValidator)
```
Comment (glidester) [inline test/controllers/MrnEntryControllerSpec.scala:65]: Do we need to reset this both before and after each test?

File: test/services/MrnDisValidatorSpec.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
+
+    reset(cdsFileUploadConnector)
+    when(cdsFileUploadConnector.getDeclarationStatus(any[MRN])(any())).thenReturn(Future.successful(defaultHttpResponse))
+  }
+
+  override protected def afterEach(): Unit = {
+    reset(cdsFileUploadConnector)
```
Comment (glidester) [inline test/services/MrnDisValidatorSpec.scala:52]: Same duplication of resetting both before and after.

**Review by glidester — APPROVED** [review]

**Comment by MaciejRewera** [comment]
@glidester Answering all your comments on resetting mocks both before and after a test case:
This is my habit I have from one of my previous workplaces. We were maintaining huge test suites for kind of integration tests. It was very common that people were not cleaning test environment after their tests and that led to problems.
So since then I prefer to run every test case in as isolated environment as possible. To clean it before getting into test and then to clean up after it's done.
I can't see any drawback to that pattern except having a few lines of code more. Even from performance perspective it costs almost nothing. I'd bet that having these `reset()` calls in every suite would cost less time to execute than to run a single suite that uses Guice Application ;-)

File: test/connectors/CdsFileUploadConnectorSpec.scala (hunk 2 above)
Comment (MaciejRewera) [reply test/connectors/CdsFileUploadConnectorSpec.scala:120]: I was thinking about it, but these two are the only ones we care about. And if I were to add more statuses, there is a question "how many statuses should I test against?".
Maybe it is reasonable to add 500 and be done with it for now?

File: app/connectors/CdsFileUploadConnector.scala [hunk 5]
```diff
@@ -43,4 +43,8 @@ class CdsFileUploadConnector @Inject()(appConfig: AppConfig, httpClient: HttpCli
         }
         notificationOpt
       }
+
+  def getDeclarationStatus(mrn: MRN)(implicit hc: HeaderCarrier): Future[HttpResponse] =
+    httpClient.GET[HttpResponse](cdsFileUploadConfig.fetchDeclarationStatusEndpoint(mrn.value))
```
Comment (udinobi) [inline app/connectors/CdsFileUploadConnector.scala:48]: Out of curiosity... how are potential exceptions thrown by HttpClient handled here?

Comment (MaciejRewera) [reply app/connectors/CdsFileUploadConnector.scala:48]: Yes

**Review by udinobi — APPROVED** [review]

**Pull request merged**